domain/secretbackend/state/types.go

```diff
@@ -0,0 +1,139 @@
+	Content string `db:"content"`
+}
+
+// SecretBackendRow represents a single joined result from secret_backend and secret_backend_config tables.
+type SecretBackendRow struct {
+	// ID is the unique identifier for the secret backend.
+	ID string `db:"uuid"`
+	// Name is the name of the secret backend.
+	Name string `db:"name"`
+	// Type is the type of the secret backend.
+	BackendType string `db:"backend_type"`
+	// TokenRotateInterval is the interval at which the token for the secret backend should be rotated.
+	TokenRotateInterval domain.NullableDuration `db:"token_rotate_interval"`
+	// ConfigName is the name of one record of the secret backend config.
+	ConfigName string `db:"config_name"`
+	// ConfigContent is the content of the secret backend config.
+	ConfigContent string `db:"config_content"`
+}
+
+// SecretBackendRows represents a slice of SecretBackendRow.
+type SecretBackendRows []SecretBackendRow
+
+func (rows SecretBackendRows) toSecretBackends() []*coresecrets.SecretBackend {
+	// Sort the rows by backend name to ensure that we group the config.
+	sort.Slice(rows, func(i, j int) bool {
+		return rows[i].Name < rows[j].Name
+	})
+	var result []*coresecrets.SecretBackend
+	var currentBackend *coresecrets.SecretBackend
+	for _, row := range rows {
+		backend := coresecrets.SecretBackend{
+			ID:          row.ID,
+			Name:        row.Name,
+			BackendType: row.BackendType,
+		}
+		interval := row.TokenRotateInterval
+		if interval.Valid {
+			backend.TokenRotateInterval = &interval.Duration
+		}
+
+		if currentBackend == nil || currentBackend.ID != backend.ID {
+			// Encountered a new backend.
+			currentBackend = &backend
+			result = append(result, currentBackend)
+		}
+		if row.ConfigName == "" || row.ConfigContent == "" {
+			// No config for this row.
+			continue
+		}
+
+		if currentBackend.Config == nil {
+			currentBackend.Config = make(map[string]interface{})
+		}
+		currentBackend.Config[row.ConfigName] = row.ConfigContent
+	}
+	return result
+}
+
+// SecretBackendRotationRow represents a single joined result from secret_backend and secret_backend_rotation tables.
+type SecretBackendRotationRow struct {
+	// ID is the unique identifier for the secret backend.
+	ID string `db:"uuid"`
+	// Name is the name of the secret backend.
+	Name string `db:"name"`
+	// NextRotationTime is the time at which the token for the secret backend should be rotated next.
+	NextRotationTime sql.NullTime `db:"next_rotation_time"`
+}
+
+type SecretBackendRotationRows []SecretBackendRotationRow
+
+func (rows SecretBackendRotationRows) toChanges(logger Logger) []watcher.SecretBackendRotateChange {
+	var result []watcher.SecretBackendRotateChange
+	for _, row := range rows {
+		change := watcher.SecretBackendRotateChange{
+			ID:   row.ID,
+			Name: row.Name,
+		}
+		next := row.NextRotationTime
+		if !next.Valid {
+			logger.Warningf("secret backend %q has no next rotation time", change.ID)
+			continue
+		}
```

Comment:
These feels like a cross-cutting concern.

I'd probably just have a validate on the `SecretBackendRotationRows` that reports the error. That way you can handle the error if you want or not.

Reply:
This field is defined as NOT NULL. So the above case should not happen(but just in case of a dev problem).
We use `sql.NullTime` more for its conversion between TEXT <->  time.Time.